### PR TITLE
Add a `CONTRIBUTING.md` document

### PR DESCRIPTION
**Pull request by voodoos**
This is a first draft of the CONTRIBUTING document in an effort to make Merlin less intimidating to potential contributors :-)

Any ~~feedback~~ contribution is welcome :-) 

cc @xvw @pitag-ha

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+issue as not already been submitted.
+
+## Code contributions
+
+ ### Legal Notice
+ When contributing to Merlin, you must agree that you have authored 100% of the
+ content, that you have the necessary rights to the content and that the content
+ you contribute may be provided under the [project's MIT
+ license](https://github.com/ocaml/merlin/blob/master/LICENSE).
+
```
Comment (sidkshatriya) [inline CONTRIBUTING.md:44]: (Disclaimer: I am *not* a lawyer)

Was this proposed Legal Notice framed by you or is it something that is being used by other projects also ?

Many contributions to ocaml merlin may modify and reuse existing ocaml merlin code, for example, so I wonder if the the stipulation of 100% authorship requirement in the proposed Legal Notice will be correct here.

Have you considered using something using the Developer Certificate of Origin 1.1 ? It is used by a lot of open source projects and may fulfill the objectives you had when you put up the proposed Legal Notice.

Comment (xvw) [reply CONTRIBUTING.md:44]: I think CLAs, in general, apply to more ambiguous license cases. In the first instance, I think the request, albeit vague (for example, how to apply it to version migrations, which vendor existing code), to be "morally the author of a patch" is sufficient.

Or maybe it's better just to delete this paragraph?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+```bash
+$ opam pin menhir 20201216
+```
+The generated parser file should only be commited if there is an actual change
+in the grammar.
+
+### Repository organization
+
+
+#### `/src/ocaml`
```
Comment (xvw) [inline CONTRIBUTING.md:73]: I think this section should be explicit about the fact that certain parts of the code (and pointed) should not be modified (because they're vendor-defined).

File: CONTRIBUTING.md (hunk 1 above)
Comment (sidkshatriya) [reply CONTRIBUTING.md:44]: This proposed "Legal Notice" seems to try to obtain some protections given by typical CLAs. However, my main objection is that the language of the "Legal Notice" feels informal / may not cover all contribution scenarios where, say, for example, some of the contribution may be a modified / improved / built-upon version of code that already exists in merlin. In that case 100% authorship required by the proposed "Legal Notice" will probably not be met by that contribution.

> "morally the author of a patch" is sufficient

Yes, agree, but whatever language is used needs to be accurate and precise, because after all, it has the heading "Legal Notice".

In conclusion, if the merlin project needs explicity outlined protections from contributions , it should use something more formal and accurate. I have suggested the Developer Certificate of Origin 1.1 as a possible solution; there are of course more ways to achieve this via proper CLAs etc.

In my opinion, more thought must put into this proposed "Legal Notice".

Comment (xvw) [reply CONTRIBUTING.md:44]: @sidkshatriya For the time being, what do you think about dropping the "Legal notice" section?

Comment (sidkshatriya) [reply CONTRIBUTING.md:44]: > @sidkshatriya For the time being, what do you think about dropping the "Legal notice" section?

I've already expressed my personal opinion above on this proposed "Legal Notice" -- I find it problematic for the reason's outined above. @voodoos has not replied to these comments so I'm not aware of their opinion on the points we have been discussing. I had suggested an alternative (Developer Certificate of Origin) too but I lack context on the motivation for this proposed change.

OCaml merlin project owners/managers should make a decision on any new requirements/stipulations for future contributions and if they want to add the proposed "Legal Notice" or not. (I am of course not part of any decision making structure on this project).

Again as a personal opinion, contributing to open source projects should be as simple and straightforward as possible. If @xvw 's suggestion on dropping the proposed "Legal Notice" keeps things simple then the project owners/managers should consider it.

Comment (voodoos) [reply CONTRIBUTING.md:44]: Thanks for having a look and raising our awareness @sidkshatriya ! I agree that the language is blurry and it probably, at best, achieve nothing in the current state. There is no deep thinking or careful writing behind the current version, it was just a generic version found in a model that I used for drafting the document. I will remove it for now until more discussions happen.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+## Question, bug reports and feature requests
+
+We rely on [Github's issue tracker](https://github.com/ocaml/merlin/issues) for
+support questions, feature requests and bug reports.
+
+When reporting an issue, please include a precise reproduction in the bug report
+when that's possible, as it is a very useful tool to investigate. You should
+also check that you are using the latest version of Merlin and that a similar
+issue as not already been submitted.
```
Comment (3Rafal) [inline CONTRIBUTING.md:35]: should be `has` instead of `as` :)

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+#### `/src/analysis`
+Contains most of Merlin logic required to answer the queries.
+
+#### `/src/dot-protocol`
+The configuration protocol that configuration servers such as
+`dot-merlin-reader` or `dune ocaml-merlin` should implement.
+
+#### `/src/dot-merlin-reader`
+The historical documentation server of Merlin. Tt reads `.merlin` configuration
```
Comment (3Rafal) [inline CONTRIBUTING.md:92]: Should be `It` instead of `Tt`

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+#### `/src/frontend/ocamlmerlin`
+The standard Merlin frontend, composed of a long-running server and a client.
+This is the frontend used by the test-suite and standard emacs and vim plugins.
+For an alternative frontend, see
+[ocaml-lsp-server](https://github.com/ocaml/ocaml-lsp/)
+
+#### `/tests/test-dirs`
+Merlin's behavioral test-suite. We rely on Dune's cram test framework. Ideally,
+any opened issue should be accompanied by a test with a reproduction. When
```
Comment (3Rafal) [inline CONTRIBUTING.md:104]: I wonder if splitting this sentence into two is more readable. Something along the lines of:
`
When working on a fix for an issue, first commit should contain the test showing the issue. Following commits should fix the issue and update the test result accordingly.
`

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+Merlin (partial and fragmented) documentation can be found in the following
+places:
+- Installation instructions and editor setup can be found in the
+  [README.md](README.md) and the [Github Pages](https://ocaml.github.io/merlin/)
+- Various topics related to editor support and configuration can be found in the
+  [wiki](https://github.com/ocaml/merlin/wiki).
+- The Merlin protocol is described in [this
+  document](https://github.com/ocaml/merlin/blob/master/doc/dev/PROTOCOL.md).
+
+Contribution to the documentation are welcome!
```
Comment (pitag-ha) [inline CONTRIBUTING.md:25]:
```suggestion
Contributions to the documentation are welcome!
```

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+version of Menhir. To remove this diff please use version `20201216`:
+```bash
+$ opam pin menhir 20201216
+```
+The generated parser file should only be commited if there is an actual change
+in the grammar.
+
+### Repository organization
+
+
```
Comment (pitag-ha) [inline CONTRIBUTING.md:65]:
```suggestion
There is a partial overview over the repo organization with file-granularity at
[ARCHITECTURE.md](https://github.com/ocaml/merlin/blob/master/doc/dev/ARCHITECTURE.md).
Here is a more high-level overview.

```

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+#### `/src/ocaml`
+Vendored OCaml typer. Also include Merlin's incremental parser and other
+utilites.
+
+#### `/src/kernel`
+The backbone of merlin. Contains configuration utilities and pipeline
+management. The pipeline describe the compilation workflow: parsing, ppxing and
+typing.
```
Comment (pitag-ha) [inline CONTRIBUTING.md:73]:
```suggestion
management. The pipeline describes the compilation workflow: parsing, ppxing and
typing.
```

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+This file contains the dispatcher: a large pattern matching that answers
+Merlin's queries. Most of the time answering a query begins with getting the
+Typedtree from the pipeline and calling the correct analysis functions.
+
+#### `/src/analysis`
+Contains most of Merlin logic required to answer the queries.
+
+#### `/src/dot-protocol`
+The configuration protocol that configuration servers such as
+`dot-merlin-reader` or `dune ocaml-merlin` should implement.
```
Comment (pitag-ha) [inline CONTRIBUTING.md:89]: Is it worth mentioning Dune RPC here as well as an alternative to `dot-merlin-reader` and `dune ocaml-merlin` (it's a genuine question)? We don't have documentation on how Dune communicates with the Merlin server via RPC that we could point to, right?

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+This is the frontend used by the test-suite and standard emacs and vim plugins.
+For an alternative frontend, see
+[ocaml-lsp-server](https://github.com/ocaml/ocaml-lsp/)
+
+#### `/tests/test-dirs`
+Merlin's behavioral test-suite. We rely on Dune's cram test framework. Ideally,
+any opened issue should be accompanied by a test with a reproduction. When
+working on a fix for an issue, please have a first commit with the test showing
+the issue, before commits that fix the issue and update the test result
+accordingly.
```
Comment (pitag-ha) [inline CONTRIBUTING.md:107]: The comment that we appreciate a test-driven approach for PRs (everything from "Ideally," on) can be missed very easily here. What do you think about moving it to the _Styleguides_ section instead of having it here in _Repository Organization_?

**Review by pitag-ha — APPROVED** [review]
Thanks a lot! This is very helpful for new contributors, and sets a clear sign that contributions are welcome!

**Review by xvw — APPROVED** [review]
